File: Algorithms/Data Structure dependent/Arrays and String/Merge Intervals/intervalIntersection.py
class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        p1, p2, e1, e2 = 0, 0, len(firstList), len(secondList)
        out = []
        while p1 != e1 and p2 != e2:
            # your ending has to be bigger than my start and my start has to be bigger than your ending
            if firstList[p1][0] <= secondList[p2][1] and firstList[p1][1] >= secondList[p2][0]:
                out.append((max(firstList[p1][0], secondList[p2][0]), min(firstList[p1][1], secondList[p2][1])))
            if firstList[p1][1] < secondList[p2][1]:
                p1 += 1
            else:
                p2 += 1

        return out



        

File: Algorithms/Data Structure dependent/Arrays and String/Merge Intervals/intervalIntersection.cpp
class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& firstList, vector<vector<int>>& secondList) {
        if(firstList.empty() || secondList.empty()) return vector<vector<int>>();
        auto p1 = firstList.begin(), p2 = secondList.begin();
        vector<vector<int>> out;
        while(p1 != firstList.end() && p2 != secondList.end()){
            auto i1 = *p1, i2 = *p2;
            if((i1[1] >= i2[0] && i1[0] <= i2[1]) || (i2[1] >= i1[0]) && i2[0] <= i1[1] ){
                vector<int> interval = {std::max(i1[0], i2[0]), std::min(i1[1],i2[1])};
                out.push_back(interval);
            }
            if(i1[1] < i2[1]) p1++;
            else p2++;

        }

        return out;
    }
};